Improve unit testing of 'password' lookup

alikins opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->

lib/ansible/plugins/lookup/password.py
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.2.0 (password_lookup ab5bd4a9eb) last updated 2016/09/08 14:30:44 (GMT -400)
  lib/ansible/modules/core: (detached HEAD db38f0c876) last updated 2016/09/07 12:50:14 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 8bfdcfcab2) last updated 2016/09/07 12:50:14 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### SUMMARY

Was attempting to reproduce https://github.com/ansible/ansible/issues/17461 with
unit tests, and these were changes to make the module a little easier to test. ( Note
this doesn't fix https://github.com/ansible/ansible/issues/17461)

The tests showed some UnicodeErrors for the
cases where the 'chars' param include unicode,
causing the 'getattr(string, c, c)' to fail.
So the candidate char generation code try/excepts
UnicodeErrors there now.

Some refactoring of the password.py module to make
it easier to test, and some new tests that cover more
of the password and salt generation.

alikins committed:
Improve unit testing of 'password' lookup

The tests showed some UnicodeErrors for the
cases where the 'chars' param include unicode,
causing the 'getattr(string, c, c)' to fail.
So the candidate char generation code try/excepts
UnicodeErrors there now.

Some refactoring of the password.py module to make
it easier to test, and some new tests that cover more
of the password and salt generation.

alikins commented:
assigned to @abadger to verify the text encoding issues handling

alikins committed:
More refactoring and fixes.

abadger commented:
- I do like breaking the run() function up.  Nice job there.
- I'd probably move those all out to toplevel private functions.... None of them use the class except to call other methods which themselves should be at the toplevel.  (Too many methods vs functions is an artifact of how Ansible was originally coded.)
- The  `test_gen_password()` unittest shouldn't call `parse_parameters()` as that is already being tested on that data in `test_parse_parameters()`.  Instead directly use old_style_params_data[FOO]['filename'] and old_style_params_data[FOO]['params'].
  - Similarly, you probably want a unittest for `get_candidate_chars()`.  Once you have that, you don't have to call it in `test_gen_password()` anymore as you'll have the chars you expect as part of the test data.
- I don't think the text encoding issues are being handled right... I actually have some code started for this because mscherer was looking at porting this to python3 and I noticed that the way characters were being selected for the password would fail horribly if non-ascii was given: https://github.com/ansible/ansible/pull/17293/files#r76843964  (which might be the same problem as  #17461 ). Let me push my branch up and you can build off of that.

abadger commented:
Here's my changes: https://github.com/ansible/ansible/pull/17475  I haven't tested it yet so there may well be bugs in there but I'm way too tired to have another looks at it tonight... will definitely be willing to go over it in IRC/slack with you tomorrow, though.

alikins committed:
manual merge of text enc fixes from pr17475

alikins committed:
moving methods to module scope

alikins committed:
more refactoring

alikins committed:
A few more text encoding fixes/merges

alikins committed:
remove now unused code

alikins committed:
Add test cases and data for _gen_candidate_chars

alikins committed:
more test coverage for password lookup

alikins committed:
wip

alikins committed:
More text encoding fixes and test coverage

alikins commented:
Took me longer to get reasonable test coverage here than expected, but should be close now.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -32,10 +32,14 @@
 
 
+# ALIKINS
+
```
Debugging stuff left in?

abadger commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -20,117 +20,349 @@
+
+from ansible.errors import AnsibleError
+from ansible.plugins import PluginLoader
+from ansible.plugins.lookup import password
```
Style-wise, I like to make the thing being tested the last import and visually separate from the other imports (a blank line between them and the import for the thing(s) being tested.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    password = []
+    while len(password) < length:
+        new_char = random_generator.choice(chars)
+        if new_char in chars:
```
Ah, this conditional is no longer needed as the new code only chooses from the valid characters.

alikins replied:
Thought that seemed unneeded. Removing.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    # Asserts would be perfect for this if they were disabled by default.
+    # Alas, in python, they're active by default; only turned off if
+    # python is run with optimization.
+    # assert isinstance(chars, text_type)
```
I've been thinking about this, I think we do need to either error or convert if the chars are not text type.  If bytes type creeps in by mistake, then random_generator.choice(chars) will pick mangled byte values.  So I suppose uncomment the assert and I can track down any bugs that it finds.

alikins replied:
Hmm, strings.digits (and ansible.constants.DEFAULT_PASSWORD_CHARS) are py2 strs on py2 and fail this assert. 

Moving the default for chars into the method and a None default arg

alikins replied:
Though, I'm a little surprised a value from constants ended up as a regular string type...

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    return chars
+
+
+def _gen_password(length, chars):
```
We don't need both `_gen_password()` and `_random_password`.  Remove one of them.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+def _gen_candidate_chars(characters):
+    chars = []
+    for chars_spec in characters:
+        chars.append(to_text(getattr(string, to_native(chars_spec), chars_spec),
```
Add a comment here that we're looking for things like "digits", "ascii_letters", etc.  Everytime I see this I spend a  little while figuring out what it does all over again.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    for chars_spec in characters:
+        chars.append(to_text(getattr(string, to_native(chars_spec), chars_spec),
+                            errors='surrogate_or_strict'))
+    chars = u''.join(to_text(c) for c in chars).replace(u'"', u'').replace(u"'", u'')
```
The to_text() and generator expression aren't needed here as you ran to_text on everything in chars above.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+
+def _gen_password(length, chars):
+    chars = _gen_candidate_chars(chars)
+    password = u''.join(random.choice(chars) for _ in range(length))
```
This uses the pseudo random number generator in the stdlib.  That's documented as being unsuited for cryptographic work.  If you keep _gen_password() instead of _random_password(), you'll need to use SystemRandom() as we do there.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    try:
+        makedirs_safe(b_pathdir, mode=0o700)
+    except OSError as e:
+        raise AnsibleError("cannot create the path for the password lookup: %s (error was %s)" % (b_pathdir, str(e)))
```
Exceptions take native strings.  So you should do this like:

``` python
raise AnsibleError("cannot create the path for the password lookup: %s (error was %s)" % (to_native(b_pathdir), to_native(e)))
```

Byte strings could cause problems on python3 and str(e) could traceback.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    password = u''.join(random.choice(chars) for _ in range(length))
+    return password
+
+def _create_password_file(b_path):
```
I notice this doesn't actually create the password file.  Perhaps "_create_password_file_dir"

Also, this is just a very light wrapper around makedirs_safe() and is called from a single location.... Perhaps it doesn't deserve its own function?

alikins replied:
The name is bad.

It is split into it's own method so that it can be easily mocked for the unit tests.

alikins replied:
I haven't figured out a good way to split _read_or_create_file better... The logic of the module is a little odd in that regards.

read_or_create_password_file() is sort of get_to_the_state_where_we_are_ready_to_write_the_password_file()

It's crammed into one method so that later code can avoid having to deal with the two potential scenarios. And, or, avoid  abstractions like

``` python

pw_store =  load_password_store()
pw_data = pw_store.read()
pw_data = update_pw_data(pw_data)
pw_store.update(pw_data)
pw_store.save()
```

abadger replied:
You can just monkeypatch makedirs_safe instead.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    except OSError as e:
+        raise AnsibleError("cannot create the path for the password lookup: %s (error was %s)" % (b_pathdir, str(e)))
+
+def _read_password_file(b_path):
```
Also a light wrapper around file.read()... Could remerge it with create_or_read_password_file.  Maybe this one has some justification because you can unittest with monkey patching this instead of mocking open if this is here... I'm not a fan of writing more complex code to make unittests easier but I can see where that might be desirable....

alikins replied:
Yup, primarily for better testing.

And to keep os/io level stuff slightly encapsulated in lieu of a more complicated abstraction (pw_store, etc).

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+
+def _read_password_file(b_path):
+    content = open(b_path, 'rb').read().rstrip()
+    return content
```
In general, if we're going to treat something as text we want to transform it to text at the border. Since we never use password as a byte string, it's just when reading and writing the file that it is bytes, we should transform it here and return text:

``` python
    b_content = open(b_path, 'rb').read().rstrip()
    return to_text(b_content)
```

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    return content
+
+def _parse_password(b_content):
+    content = to_text(b_content, errors='surrogate_or_strict')
```
made a note earlier about transforming this to text when we read it from the file.  So we can pass in content as text and not do a transformation here.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+def _write_password_file(b_path, content):
+    with open(b_path, 'wb') as f:
+        os.chmod(b_path, 0o600)
+        f.write(content)
```
Just like transforming into text as soon as possible, we also want to transform into bytes as late as possible.  So we want to do the to_bytes() call here rather than before _write_password_file() is called:

``` python
with open(b_path, 'wb') as f:
    os.chmod(b_path, 0o600)
    f.write(to_bytes(content))
```

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
-                password = do_encrypt(password, params['encrypt'], salt=salt)
+            # save it
+            b_content = to_bytes(content, errors='surrogate_or_strict') + b'\n'
+            _write_password_file(b_path, b_content)
```
As noted earlier, pass the content as a text string and let _write_password_file transform it into bytes.  So:

``` python
content = content + u'\n'
_write_password_file(b_path, content)
```

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
-    def run(self, terms, variables, **kwargs):
+    if os.path.exists(b_path):
+        b_file_content = _read_password_file(b_path)
+        content, plaintext_password, salt = _parse_password(b_file_content)
```
As noted earlier, we'd be passing this around as file_content (text), rather than b_file_content (bytes).

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+    else:
+        _create_password_file(b_path)
+
+    return content, plaintext_password, salt
```
The split in responsibilities between this and _update_content feels wrong.  Here we're responsible for reading the information from the file and returning it along with the parsed form.  But if we don't have a file, we return data that flags an this step as incomplete.  In _update_content, we're responsible for producing a new combined content from the parsed form.  But if we get the flag, we have to create the exploded form of the data from scratch.

Instead we should make it this function's responsibility to return the parsed data.  It will either read the data in or create the data.  _update_content() can then go away.... We may want to call _format_content() in its place or (since we're returning the content here) we may want to call _format_content() from this function as well.

alikins replied:
I can add a wrapper function that will call both, but I'd prefer to keep the impls as separate methods.

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
+
+    return content, plaintext_password, salt
+
+def _update_content(plaintext_password, salt, params):
```
As noted above, some or all of this should be merged with _read_or_create_password_file().

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +77,149 @@ def _parse_parameters(term):
-                        os.chmod(path, 0o600)
-                        f.write(content + '\n')
+            content, plaintext_password, salt = _read_or_create_password_file(b_path)
+            content, plaintext_password, salt = _update_content(plaintext_password, salt, params)
```
As noted earlier, _update_content should be merged with _read_or_create_password_file().  You may end up calling _format_content here in its place or that might be merged into _read_or_create_password_file().

abadger commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -20,117 +20,349 @@
+        results = password_lookup.run([u'/path/to/somewhere encrypt=pbkdf2_sha256'],
+                                      None)
+
+        # FIXME: assert something useful
```
For encrypted you should be able to assert something about the format:

```
>>> crypt = getattr(passlib.hash, 'pbkdf2_sha256')
>>> crypt.encrypt(u'asda')
'$pbkdf2-sha256$20000$PUeoVar1/h8jxNjb./8fgw$UzKj5pfmKEDtguOh.XEvHCqQr2b4AHPgwL/M8BEzez8'
```

Initial string, number of "$", length of the salt and length of the hashed value.

abadger commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -20,117 +20,349 @@
+                                  None)
+        for result in results:
+            for char in result:
+                self.assertEqual(char, u'a')
```
You should be able to know the length too (either the default or specify it) and then you don't have to loop through each character in result.

abadger commented on test/units/plugins/lookup/test_password.py:
```diff
@@ -20,117 +20,349 @@
+
+class TestParseParamaters(unittest.TestCase):
+    def test(self):
+        for testcase in old_style_params_data:
```
Could make this into a generated testcase if you were so inclined.

alikins committed:
cleanups

alikins committed:
reenable text_type assert

alikins committed:
Remove unneeded conditional in _random_password

alikins committed:
Add docstring for _gen_candidate_chars

alikins committed:
remove redundant to_text and list comphenesion

alikins committed:
Move set of 'chars' default in _random_password

on py2, C.DEFAULT_PASSWORD_CHARS is a regular str
type, so the assert here fails. Move setting the
default into the method and to_text(DEFAULT_PASSWORD_CHARS)
if it's needed.

alikins committed:
combine _random_password and _gen_password

alikins committed:
s/_create_password_file/_create_password_file_dir

alikins committed:
native strings for exception msgs

alikins committed:
move password to_text to _read_password_file

alikins committed:
move to_bytes(content) to _write_password_file

alikins committed:
add more test assertions about genned pw's

alikins commented:
Think i have all the changed done.

Sans the flow around _read_or_create_password_file_dir().

abadger commented on lib/ansible/plugins/lookup/password.py:
```diff
@@ -73,91 +75,163 @@ def _parse_parameters(term):
+    # Asserts would be perfect for this if they were disabled by default.
+    # Alas, in python, they're active by default; only turned off if
+    # python is run with optimization.
+    chars = chars or to_text(C.DEFAULT_PASSWORD_CHARS)
```
Convert C.DEFAULT_PASSWORD_CHARS into text in lib/ansible/constants,py instead.

abadger commented:
These commits and additional changes merged via: https://github.com/ansible/ansible/pull/17626
